src/diffusers/loaders.py

```diff
@@ -1200,16 +1217,25 @@ def save_function(weights, filename):
             state_dict.update(unet_lora_state_dict)
 
         if text_encoder_lora_layers is not None:
-            weights = (
-                text_encoder_lora_layers.state_dict()
-                if isinstance(text_encoder_lora_layers, torch.nn.Module)
-                else text_encoder_lora_layers
-            )
 
-            text_encoder_lora_state_dict = {
-                f"{self.text_encoder_name}.{module_name}": param for module_name, param in weights.items()
-            }
-            state_dict.update(text_encoder_lora_state_dict)
+            def get_te_params(te_state_dict):
+                weights = te_state_dict.state_dict() if isinstance(te_state_dict, torch.nn.Module) else te_state_dict
+                return weights
+
+            if not isinstance(text_encoder_lora_layers, list):
+                weights = get_te_params(text_encoder_lora_layers)
+                text_encoder_lora_state_dict = {
+                    f"{self.text_encoder_name}.{module_name}": param for module_name, param in weights.items()
+                }
+                state_dict.update(text_encoder_lora_state_dict)
+            else:
+                for i in range(len(text_encoder_lora_layers)):
+                    weights = get_te_params(text_encoder_lora_layers[i])
+                    weight_suffix = f"{self.text_encoder_name}_{i+1}" if i > 0 else self.text_encoder_name
+                    text_encoder_lora_state_dict = {
+                        f"{weight_suffix}.{module_name}": param for module_name, param in weights.items()
+                    }
+                    state_dict.update(text_encoder_lora_state_dict)
```

Comment:
I think this is quite implicit and relies on a naming convention in the sub class, I think it would be better if we just overrode the method in the sdxl pipeline and moved the file writing code to a separate classmethod

i.e.

```py
class LoraLoaderMixin:
    @classmethod
    def save_lora_weights(...):
        state_dict = ...
        self.write_lora_weights(state_dict, ...)

   @classmethod
   def write_lora_weights(state_dict):
       ...
        
class SDXLPipeline:
    @classmethod
    def save_lora_weights(..., unet_lora_layers, text_encoder_lora_layers, text_encoder_2_lora_layers):
         state_dict = {}
         state_dict.update({f"unet.{k}": v for k, v in unet_lora_layers.state_dict()})
         state_dict.update({f"text_encoder.{k}": v for k, v in text_encoder_lora_layers.state_dict()})
         state_dict.update({f"text_encoder_2.{k}": v for k, v in text_encoder_lora_layers.state_dict()})

         self.write_lora_weights(state_dict, ...)
```

A good rule of thumb is that we should ideally treat the state dict as opaque as possible -- i.e. we just write it and load from it. Any time we have to manually either look inside it or choose what keys to write to it, be as explicit as possible.

Reply:
Ah, I think we can still keep the changes if we introduced a `prefix` argument in `save_lora_weights()` function following https://github.com/huggingface/diffusers/pull/4097/files#r1266022763? I think that still respects what you're trying to achieve here. 

@patrickvonplaten, would love to hear your thoughts too.

Reply:
Hmm yeah this one is a bit tricky. Tend to agree with @williamberman here that creating a new function seems to be most sensible here given that SD-XL is arguably a big exception since it has two text encoders.

Reply:
If we go via this way, then we will need to initialize the SDXL pipeline in the training script just for serialization purposes. This also applies to the hooks we use for checkpointing. Are we okay with that?

Reply:
I don't see what requires instantiating an instance of the class, they're all classmethods. I'm also not really sure I see how this impacts tests unfortunately. 

@sayakpaul could you show the pseudocode for how passing a prefix fixes it?

Reply:
Ok I see the existing update that adds the prefix. I'm not sure I really see the difference between as it was before with adding the `text_encoder_prefix` argument, the body of the function is almost the same just the value is in a function argument? We still have the same implicit contract that the sub class must name its text encoders a particular way.

I feel relatively strongly that we should do as recommended here https://github.com/huggingface/diffusers/pull/4097#discussion_r1266035100